Frontend/update graph boundries#213

lposo opened the pull request:
Updated the graph page so that boundaries can be modified in increments of 1. The boundaries of an axis must contain some points, else it cannot be changed.

Screenshots:

This image shows the graph  with the x-upper limit at its minimum value.
![image](https://user-images.githubusercontent.com/46732530/104972938-39e80800-59c1-11eb-940d-529856d5b538.png)

This image shows the graph with the upper limits at their minimal values and the lower limits at their maximum values.
![image](https://user-images.githubusercontent.com/46732530/104973013-787dc280-59c1-11eb-9c2f-88cf1c20702b.png)

lposo committed:
Added ability to change boundaries for Graph.

Users can now modify the boundaries for the x and y axes of the graph.

lposo committed:
Finished implementation of X and Y boundaries.

Users can now specify the boundaries for the x and y axis. Additionally,the boundaries must contain some data, else they cannot be changed.

lposo committed:
Merge branch 'develop' into Frontend/UpdateGraphBoundries#213

lposo committed:
Merge branch 'develop' into Frontend/UpdateGraphBoundries#213

TiraFire reviewed (changes requested):
First thing is this does not match the mockup. There is a link in your story that references the mockup in #212 and shows how this is an addition to the settings.

When you make boundaries, you should not be able to move the graph outside the bound. Since this is not the case, a bug occurs when you have the graph outside the bounds but zooming out is no longer functional because the bounds only account for the difference between the upper and lower. For instance:
![image](https://user-images.githubusercontent.com/25015601/104979384-26916880-59d2-11eb-954e-9927fd15ab2b.png)
The x upper and lower set by the user is 1 and 8 but if you move the graph, you can change it as shown in the screenshot where it still has the difference of 7 (0-7)

Also, there seems to be some bugs : There are moments I cannot increase my upper boundaries. To replicate this error, go to each of the limits of a dataset, then try increasing the Y upper bound. 

![image](https://user-images.githubusercontent.com/25015601/104979159-97845080-59d1-11eb-89d1-402c04a4d789.png)

Also, it should have at least 1 data point, not 1 data set

Ill continue some testing but these should be fixed

ds161833 commented on Client/src/Components/Graph/Graph.tsx:
```diff
@@ -34,16 +34,86 @@ export default function Graph(props: IProps) {
 
   const [isXLog, setXToggle] = useState(false)
   const [isYLog, setYToggle] = useState(false)
+  const [minX, setMinX] = useState(9999)
```
use one object to define all this, also, are Min/Max values really needed? whats the problem with a user setting the bound too far, also here is a nice chart from another lib that has that feature embedded: https://www.amcharts.com/demos/dumbbell-plot/

TiraFire replied:
I guess overall, we can just give the tools to the user, and let them do what they want. But if the boundaries are there, i still want them respected. So no zooming out further than the bounds or dragging it out of bounds. The only different with your example is the lack of the "drag" feature directly on the graph itself

lposo committed:
Fixed boundaries sometiems not being changable and fixed being able to zoom past boundaries.

Boundaries should now be changable at all times. Additionally, zooming and panning should only work up to the specified boundaries.

lposo committed:
Fixed issue with adding a additional datasets when boundaries are at their extremes.

After adding another dataset whose extreme values are beyond the currently set boundaries, attempting to modify said boundary will instead set it to the new max/min.

lposo committed:
Moved calculationh of graph extremities to GraphView

This portion of the code was moved because of a bug that occured when removing a graph that had more extreme points than a previous dataset. Before, the max/min would remain as the max/min of the previous dataset/ Now, as it is calculated whenever changes to the loaded dataset are done, this issue should no longer exist.

lposo committed:
Added button to toggle/untoggle settings.

lposo committed:
Added icon to button.

lposo commented:
![image](https://user-images.githubusercontent.com/46732530/105107533-58fc9d80-5a86-11eb-80eb-82618f14f451.png)
New toggle for settings. Additionally, you can no longer zoom/pan past the boundaries that are set.

ds161833 commented on Client/src/Components/Graph/Graph.tsx:
```diff
@@ -31,19 +35,74 @@ export default function Graph(props: IProps) {
 
   const handleXScaleClick = () => {
+    if (xLowerBound <= 0 && !isXLog) {
+      setXLowerBound(1)
```
you can use the whole object as a state, that way you only have one state representing the boundaries, you then set it with: setBoundaries({...boundaries, minY: 1 }) etc.

lposo replied:
I've modified it to now use an object as a state.

ds161833 commented on Client/src/Components/Graph/GraphView.tsx:
```diff
@@ -77,24 +78,52 @@ export default function GraphView() {
   }
 
+  function calculateExtremeBoundaries(datasets: ICompleteDatasetEntity[]) {
+    let minX = 9000, maxX = 0, minY = 9000, maxY = 0
```
isnt it possible to have negative values for Min/Max?

lposo replied:
Yes, those are just temporary values. Within the actual graph, you can put negative values as long as you are not in the logarithmic scale for that axis.

ds161833 commented on Client/src/Components/Graph/Graph.tsx:
```diff
@@ -31,19 +35,74 @@ export default function Graph(props: IProps) {
+    }
+  }
+
+  const handleYUpperBoundChange = (event) => {
```
you could actually have this all done as a formik form and then you wouldnt need to write your own handlers

lposo replied:
I'll try to modify it to use Formik instead.

lposo replied:
I decided not to use Formik due to it working oddly when trying to override the FastField with an onchange prop.

ds161833 commented on Client/src/Components/Graph/Graph.tsx:
```diff
@@ -31,19 +35,74 @@ export default function Graph(props: IProps) {
+      setYUpperBound(Math.ceil(extremeBoundaries.maxY))
+    }
+    else {
+      setYUpperBound(value)
```
https://www.tutorialspoint.com/How-to-limit-the-number-of-characters-allowed-in-form-input-text-field#:~:text=To%20give%20a%20limit%20to%20the%20input%20field%2C%20use%20the,characters%2C%20use%20the%20maxlength%20attribute. try looking if setting a character limit on an input would be easier then manualy checking the boundary

lposo replied:
The limit is more in place so that the entire dataset is always within the boundaries specified. I don't know if setting a character limit will really help make it simpler.

lposo committed:
Moved boundaries to an object state instead of each being its own individual state.

lposo committed:
Merge branch 'develop' into frontend/UpdateGraphBoundries#213

ds161833 commented:
good for now but im afraid that if we change the graphing lib it may come there out of the box

GD-Dheer committed:
Merge branch 'develop' into frontend/UpdateGraphBoundries#213
